lib/BC/SleighLifter.cpp

```diff
@@ -1279,14 +1364,26 @@ std::map<OpCode, SleighLifter::PcodeToLLVMEmitIntoBlock::BinaryOperator>
            if (lhs->getType() != rhs->getType()) {
              rhs = bldr.CreateZExtOrTrunc(rhs, lhs->getType());
            }
-           return bldr.CreateShl(lhs, rhs);
+           // If the number of bits we're shifting exceeds the bit width of the
+           // other operand, the result should be zero.
+           auto max_shift = llvm::ConstantInt::get(
+               lhs->getType(), lhs->getType()->getIntegerBitWidth());
+           return bldr.CreateSelect(bldr.CreateICmpSGE(rhs, max_shift),
+                                    llvm::ConstantInt::get(lhs->getType(), 0),
+                                    bldr.CreateShl(lhs, rhs));
          }},
         {OpCode::CPUI_INT_RIGHT,
          [](llvm::Value *lhs, llvm::Value *rhs, llvm::IRBuilder<> &bldr) {
            if (lhs->getType() != rhs->getType()) {
              rhs = bldr.CreateZExtOrTrunc(rhs, lhs->getType());
            }
-           return bldr.CreateLShr(lhs, rhs);
+           // If the number of bits we're shifting exceeds the bit width of the
+           // other operand, the result should be zero.
+           auto max_shift = llvm::ConstantInt::get(
+               lhs->getType(), lhs->getType()->getIntegerBitWidth());
+           return bldr.CreateSelect(bldr.CreateICmpSGE(rhs, max_shift),
+                                    llvm::ConstantInt::get(lhs->getType(), 0),
+                                    bldr.CreateLShr(lhs, rhs));
```

Comment:
Copy & paste of the change above. Extract.

Reply:
[9cec23d](https://github.com/lifting-bits/remill/pull/645/commits/9cec23d73bc4f9a228be3f92e3ee275141cd2987)